---
title: "MicrobiomeR: Data Wrangling"
author: "Robert Gilmore"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Data Wrangling}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(MicrobiomeR)
pkg.data <- MicrobiomeR:::pkg.private

```

# Import
Importing data is done via the phyloseq package.  MicrobiomeR was developed by testing data from [Nephele's
16S Qiime](https://nephele.niaid.nih.gov/user_guide_pipes/#amplicon_pipes) pipeline, so that will be our 
frame of reference for these vignettes.  While one can use any phyloseq object generated by phyloseq, the 
function `MicrobiomeR::get_phyloseq_obj()` will be used to import the Nephele data, which includes a [biom
file](http://biom-format.org/), a phylogenetic tree file, and the metadata file ([Nephele mapping 
file](https://nephele.niaid.nih.gov/user_guide/).
file.


(_**Note**_:  *Test data can be found in a private environment variable called with MicrobiomeR:::pkg.private*)

```{r message=FALSE, warning=FALSE}
# Get the data files from package
input_files <- pkg.data$input_files
biom_file <- input_files$biom_files$silva # Path to silva biom file
tree_file <- input_files$tree_files$silva # Path to silva tree file
metadata_file <- input_files$metadata # Path to Nephele metadata
parse_func <- parse_taxonomy_silva_128 # A custom phyloseq parsing function for silva annotations

# Get the phyloseq object
phy_obj <- get_phyloseq_obj(biom_file = biom_file, 
                            tree_file = tree_file, 
                            metadata_file = metadata_file,
                            parse_func = parse_func)
```

# Formatting
MicrobiomeR provides an intuitive formatting system for taxmap objects.  

# Visualization
```{r message=FALSE, warning=FALSE}
mc_obj <- as_analyzed_format(obj = raw_silva)
hts <- get_heat_tree_plots(mc_obj, rank_list = c("Class"))
hts$heat_trees$Class
```

