<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data Wrangling • MicrobiomeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Data Wrangling">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeR</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/data-wrangling.html">Data Wrangling</a>
    </li>
    <li>
      <a href="../articles/filtering.html">Data Filtering</a>
    </li>
    <li>
      <a href="../articles/analysis.html">Data Analysis</a>
    </li>
    <li>
      <a href="../articles/color-palettes.html">Color Palettes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/about.html">About</a>
</li>
<li>
  <a href="https://twitter.com/datasnakes">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/vallenderlab/MicrobiomeR">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Data Wrangling</h1>
                        <h4 class="author">Robert Gilmore</h4>
            
            <h4 class="date">2019-02-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vallenderlab/MicrobiomeR/blob/master/vignettes/data-wrangling.Rmd"><code>vignettes/data-wrangling.Rmd</code></a></small>
      <div class="hidden name"><code>data-wrangling.Rmd</code></div>

    </div>

    
    
<div id="import" class="section level1">
<h1 class="hasAnchor">
<a href="#import" class="anchor"></a>Import</h1>
<p>Importing data is done via the <a href="https://github.com/joey711/phyloseq">phyloseq</a> package. <a href="https://github.com/vallenderlab/MicrobiomeR">MicrobiomeR</a> was developed by testing data from <a href="https://nephele.niaid.nih.gov/user_guide_pipes/#amplicon_pipes">Nephele’s 16S Qiime</a> pipeline, which will influences the workflow for these vignettes. While one can use any phyloseq object generated by phyloseq for MicrobiomeR, we suggest using the function, <code><a href="../reference/create_phyloseq.html">MicrobiomeR::create_phyloseq()</a></code>, and we will demonstrate using it to import the Nephele data, which includes a <a href="http://biom-format.org/">biom file</a>, a phylogenetic tree file, and the metadata file (<a href="https://nephele.niaid.nih.gov/user_guide/">Nephele mapping file</a>). file.</p>
<p>(<em><strong>Note</strong></em>: <em>Test data can be found in a private environment variable called with MicrobiomeR:::pkg.private</em>)</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># Get the data files from package</span></a>
<a class="sourceLine" id="cb1-2" title="2">input_files &lt;-<span class="st"> </span>pkg.data<span class="op">$</span>input_files</a>
<a class="sourceLine" id="cb1-3" title="3">biom_file &lt;-<span class="st"> </span>input_files<span class="op">$</span>biom_files<span class="op">$</span>silva <span class="co"># Path to silva biom file</span></a>
<a class="sourceLine" id="cb1-4" title="4">tree_file &lt;-<span class="st"> </span>input_files<span class="op">$</span>tree_files<span class="op">$</span>silva <span class="co"># Path to silva tree file</span></a>
<a class="sourceLine" id="cb1-5" title="5">metadata_file &lt;-<span class="st"> </span>input_files<span class="op">$</span>metadata<span class="op">$</span>two_groups <span class="co"># Path to Nephele metadata</span></a>
<a class="sourceLine" id="cb1-6" title="6">parse_func &lt;-<span class="st"> </span>parse_taxonomy_silva_<span class="dv">128</span> <span class="co"># A custom phyloseq parsing function for silva annotations</span></a>
<a class="sourceLine" id="cb1-7" title="7"></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="co"># Get the phyloseq object</span></a>
<a class="sourceLine" id="cb1-9" title="9">phyloseq_object &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_phyloseq.html">create_phyloseq</a></span>(</a>
<a class="sourceLine" id="cb1-10" title="10">  <span class="dt">biom_file =</span> biom_file,</a>
<a class="sourceLine" id="cb1-11" title="11">  <span class="dt">tree_file =</span> tree_file,</a>
<a class="sourceLine" id="cb1-12" title="12">  <span class="dt">metadata_file =</span> metadata_file,</a>
<a class="sourceLine" id="cb1-13" title="13">  <span class="dt">parse_func =</span> parse_func</a>
<a class="sourceLine" id="cb1-14" title="14">)</a></code></pre></div>
</div>
<div id="formatting" class="section level1">
<h1 class="hasAnchor">
<a href="#formatting" class="anchor"></a>Formatting</h1>
<p>After importing amplicon data with phyloseq, you will need to convert the phyloseq object to a <a href="https://github.com/ropensci/taxa">taxa::taxmap</a> object for use with the <a href="https://github.com/grunwaldlab/metacoder">metacoder</a> package. MicrobiomeR has added additional “formatting” to the taxmap object in the form of specific tables names. For more information please consult the Formatting help page in your R console via <code><a href="../reference/MicrobiomeR_Formats.html">?MicrobiomeR::MicrobiomeR_Formats</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">p_obj &lt;-<span class="st"> </span>phyloseq_object</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="co"># Create the various formats</span></a>
<a class="sourceLine" id="cb2-3" title="3">phy_format &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_MicrobiomeR_format.html">as_MicrobiomeR_format</a></span>(<span class="dt">obj =</span> p_obj, <span class="dt">format =</span> <span class="st">"phyloseq_format"</span>)</a>
<a class="sourceLine" id="cb2-4" title="4">raw_format &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_MicrobiomeR_format.html">as_MicrobiomeR_format</a></span>(<span class="dt">obj =</span> p_obj, <span class="dt">format =</span> <span class="st">"raw_format"</span>)</a>
<a class="sourceLine" id="cb2-5" title="5">basic_format &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_MicrobiomeR_format.html">as_MicrobiomeR_format</a></span>(<span class="dt">obj =</span> p_obj, <span class="dt">format =</span> <span class="st">"basic_format"</span>)</a>
<a class="sourceLine" id="cb2-6" title="6">analyzed_format &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_MicrobiomeR_format.html">as_MicrobiomeR_format</a></span>(<span class="dt">obj =</span> p_obj, <span class="dt">format =</span> <span class="st">"analyzed_format"</span>)</a></code></pre></div>
<p>The various formats that have been defined include the “phyloseq”, “raw”, “basic”, and “analyzed” formats. These formats define the level or stage that the taxmap object is in. The only difference between MicrobiomeR taxmap objects and regular taxmap objects is the naming convention used for the observation data (e.g. <code><a href="https://www.rdocumentation.org/packages/base/topics/names">names(MicrobiomeR_obj$data)</a></code>).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># Show the difference in MicrobiomeR formats</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(phy_format<span class="op">$</span>data))</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="co">#&gt; [1] "sample_data" "phy_tree"    "otu_table"   "tax_data"</span></a>
<a class="sourceLine" id="cb3-4" title="4"></a>
<a class="sourceLine" id="cb3-5" title="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(raw_format<span class="op">$</span>data))</a>
<a class="sourceLine" id="cb3-6" title="6"><span class="co">#&gt; [1] "otu_abundance"   "otu_annotations" "sample_data"     "phy_tree"</span></a>
<a class="sourceLine" id="cb3-7" title="7"></a>
<a class="sourceLine" id="cb3-8" title="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(basic_format<span class="op">$</span>data))</a>
<a class="sourceLine" id="cb3-9" title="9"><span class="co">#&gt; [1] "otu_abundance"    "otu_annotations"  "otu_proportions" </span></a>
<a class="sourceLine" id="cb3-10" title="10"><span class="co">#&gt; [4] "sample_data"      "phy_tree"         "taxa_abundance"  </span></a>
<a class="sourceLine" id="cb3-11" title="11"><span class="co">#&gt; [7] "taxa_proportions"</span></a>
<a class="sourceLine" id="cb3-12" title="12"></a>
<a class="sourceLine" id="cb3-13" title="13"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(analyzed_format<span class="op">$</span>data))</a>
<a class="sourceLine" id="cb3-14" title="14"><span class="co">#&gt; [1] "otu_abundance"    "otu_annotations"  "otu_proportions" </span></a>
<a class="sourceLine" id="cb3-15" title="15"><span class="co">#&gt; [4] "sample_data"      "phy_tree"         "taxa_abundance"  </span></a>
<a class="sourceLine" id="cb3-16" title="16"><span class="co">#&gt; [7] "taxa_proportions" "statistical_data" "stats_tax_data"</span></a></code></pre></div>
<p>MicrobiomeR provides a group of helpful functions that can be used to format taxmap objects including:</p>
<ul>
<li>
<code><a href="../reference/which_format.html">which_format()</a></code> for returning the format of the taxmap object.</li>
<li>
<code>is_*_format()</code> for testing if an object is in a specific format.</li>
<li>
<code>as_*_format()</code> for converting an object to a specific format.</li>
<li>
<code><a href="../reference/order_metacoder_data.html">order_metacoder_data()</a></code> for ordering the observation data for analysis.</li>
<li>
<code><a href="../reference/validate_MicrobiomeR_format.html">validate_MicrobiomeR_format()</a></code> for validating that the taxmap object is in a specific format.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co"># Determine format</span></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw"><a href="../reference/which_format.html">which_format</a></span>(analyzed_format)</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="co">#&gt; [1] "analyzed_format"</span></a>
<a class="sourceLine" id="cb4-4" title="4"></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="kw"><a href="../reference/is_analyzed_format.html">is_analyzed_format</a></span>(analyzed_format)</a>
<a class="sourceLine" id="cb4-6" title="6"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb4-7" title="7"></a>
<a class="sourceLine" id="cb4-8" title="8"><span class="co"># Forcing validation</span></a>
<a class="sourceLine" id="cb4-9" title="9">low_level_format &lt;-<span class="st"> </span><span class="kw"><a href="../reference/validate_MicrobiomeR_format.html">validate_MicrobiomeR_format</a></span>(<span class="dt">obj =</span> raw_format, <span class="dt">valid_formats =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"analyzed_format"</span>, <span class="st">"basic_format"</span>), <span class="dt">force_format =</span> <span class="ot">TRUE</span>, <span class="dt">min_or_max =</span> min)</a>
<a class="sourceLine" id="cb4-10" title="10"><span class="kw"><a href="../reference/which_format.html">which_format</a></span>(low_level_format)</a>
<a class="sourceLine" id="cb4-11" title="11"><span class="co">#&gt; [1] "basic_format"</span></a>
<a class="sourceLine" id="cb4-12" title="12"></a>
<a class="sourceLine" id="cb4-13" title="13">high_level_format &lt;-<span class="st"> </span><span class="kw"><a href="../reference/validate_MicrobiomeR_format.html">validate_MicrobiomeR_format</a></span>(<span class="dt">obj =</span> raw_format, <span class="dt">valid_formats =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"analyzed_format"</span>, <span class="st">"basic_format"</span>), <span class="dt">force_format =</span> <span class="ot">TRUE</span>, <span class="dt">min_or_max =</span> max)</a>
<a class="sourceLine" id="cb4-14" title="14"><span class="kw"><a href="../reference/which_format.html">which_format</a></span>(high_level_format)</a>
<a class="sourceLine" id="cb4-15" title="15"><span class="co">#&gt; [1] "analyzed_format"</span></a></code></pre></div>
<p>However, most of the time you will simply use <code><a href="../reference/as_MicrobiomeR_format.html">as_MicrobiomeR_format(format=...)</a></code> to get an object into the proper format.</p>
<p>(<strong>Please note that MicrobiomeR <em>expects</em> the data to be imported with phyloseq and then converted to a taxmap object</strong>)</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#import">Import</a></li>
      <li><a href="#formatting">Formatting</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/grabear">Robert Gilmore</a>, <a href="https://www.shauritahutchins.com">Shaurita Hutchins</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
