<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MicrobiomeR: An R Package for Simplified and Standardized Microbiome Analysis Workflows (Draft to JOSS) • MicrobiomeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="MicrobiomeR: An R Package for Simplified and Standardized Microbiome Analysis Workflows (Draft to JOSS)">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeR</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/data-wrangling.html">Data Wrangling</a>
    </li>
    <li>
      <a href="../articles/filtering.html">Data Filtering</a>
    </li>
    <li>
      <a href="../articles/analysis.html">Data Analysis</a>
    </li>
    <li>
      <a href="../articles/color-palettes.html">Color Palettes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/paper.html">Publication</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../articles/about.html">About</a>
</li>
<li>
  <a href="https://twitter.com/datasnakes">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/vallenderlab/MicrobiomeR">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MicrobiomeR: An R Package for Simplified and Standardized Microbiome Analysis Workflows (Draft to JOSS)</h1>
                        <h4 class="author">Robert A Gilmore</h4>
                        <h4 class="author">Shaurita Hutchins</h4>
                        <h4 class="author">Xiao Zhang</h4>
                        <h4 class="author">Eric Vallender</h4>
            
            <h4 class="date">2019-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vallenderlab/MicrobiomeR/blob/master/vignettes/paper.Rmd"><code>vignettes/paper.Rmd</code></a></small>
      <div class="hidden name"><code>paper.Rmd</code></div>

    </div>

    
    
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<p>The term “microbiome” refers to the microbial community in a given environment. In particular, it has recently risen to prominence in medicine to refer to the human oral, skin, urogenital, and digestive tract commensal bacterial communities. Modern technologies allow for the characterization of microbiome composition by high-throughput sequencing of 16S ribosomal DNA that effectively serves as identifying “bar codes”. Research studies focus on comparing differences in microbiomes between environments or changes within environments over time. To ensure study rigor and reproducibility it is important that data processing be clearly described and standardized. This package aims to unify the processing, analysis, and visualization tools necessary for modern microbiome studies under a transparent and straightforward implementation that facilitates standardization in analysis and reporting.</p>
<p>Previous open source microbiome packages available for R <span class="citation">(R Core Team 2018)</span> (e.g. <strong>vegan</strong> and <strong>microbiome</strong>) have been developed, but lack the functionality afforded to more modern tools including <strong>phyloseq</strong> <span class="citation">(McMurdie and Holmes 2013)</span> and <strong>metacoder</strong> <span class="citation">(Foster, Sharpton, and Grünwald 2017)</span>. Both of these packages, however, provide different degrees of functionality as it relates to data wrangling, statistical methods, and visualization. Phyloseq, for instance, relies on <strong>base</strong> R functions such as <em>subset</em> to extract or manipulate data, while metacoder uses a more modern approach like the <strong>tidyverse</strong>. Additionally, metacoder is built on top of the <strong>taxa</strong> package and uses a <em>“taxmap object”</em>, which allows for direct manipulation of hierarchical taxonomic data and associated application-specific data <span class="citation">(Zachary, Scott, and Niklaus 2018)</span>. Phyloseq, on the other hand, provides an excellent means for importing data into R as a <em>“phyloseq object”</em>, which can be used with various proven methods for analysis. In order to bridge the gap, we have developed <strong>MicrobiomeR</strong>, to provide new tools and a comprehensive workflow based on concepts found in the phyloseq package and newer technologies being developed in the metacoder package.</p>
</div>
<div id="workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow" class="anchor"></a>Workflow</h1>
<p>The paradigm for studying the microbiome follows the steps shown in the figure below. MicrobiomeR covers the last four steps of this workflow. <img src="Microbiome-Workflow.png" alt="Microbiome Workflow"></p>
<p>MicrobiomeR provides users with raw output files from the NIH’s Nephele pipeline. This data was generated by the Illumina MiSeq workflow followed by the <strong>Qiime</strong> <span class="citation">(JG et al. 2010)</span> 16S pipeline for pair ended FASTQ files. To begin a MicrobiomeR workflow, raw data files are imported into R as a phyloseq object, using the <em>create_phyloseq()</em> function. While discouraged, the data can be manipulated directly using any of phyloseq’s methods. Next, the data is converted to a taxmap object using the <em>create_taxmap()</em> function.</p>
<p>By conceptually modeling MicrobiomeR’s data preprocessing functions after phyloseq, we are able to enhance the useful microbiome analysis capabilities of metacoder and taxa by providing simple “phyloseq-style” preprocessing functions, which are critical for reducing noise within the data. Many of the **_filter()* functions are used for this purpose. Additionally, any of the metacoder or taxa functions can be incorporated into the preprocessing steps as MicrobiomeR primarily operates on taxmap objects.</p>
<p>After preprocessing the data, statistical analysis can be conducted using MicrobiomeR’s formatting functions (<em>as_*<em>format()<em>), which incorporate metacoder’s </em>calc</em>*()</em> and <em>compare_groups()</em> functions. Furthermore, MicrobiomeR offers a unique <em>permanova()</em> function, which quantifies multivariate community-level differences between groups. Following statistical analysis, alpha and beta diversity can be visualized using the <em>alpha_diversity_plot()</em> and <em>ordination_plot()</em> functions, which are helpful for understanding the intra-sample differences (evenness and richness) and the inter-group differences. Other publication-ready visualizations can also be created using <em>heat_tree_plots()</em>, <em>correlation_plots()</em>, <em>stacked_barplot()</em>, and <em>top_coefficients_barplot()</em>.</p>
</div>
<div id="utilities" class="section level1">
<h1 class="hasAnchor">
<a href="#utilities" class="anchor"></a>Utilities</h1>
<p>One of the key advantages in using MicrobiomeR is its “phyloseq-style” filtering functions. These functions are heavily dependent upon formatting and validation checkpoints to ensure that the proper data is being manipulated. In Table 1, we describe the “MicrobiomeR formats” used throughout the package and how they are related to phyloseq, metacoder, and each other.</p>
<div id="table-1" class="section level3">
<h3 class="hasAnchor">
<a href="#table-1" class="anchor"></a>Table 1:</h3>
<table class="table">
<colgroup>
<col width="2%">
<col width="4%">
<col width="29%">
<col width="16%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th align="center">Level</th>
<th align="center">Name</th>
<th align="center">Description</th>
<th align="center">Data Tables</th>
<th align="center">Other Notes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">0</td>
<td align="center">PhyloseqFormat</td>
<td align="center">A taxmap object that has just been converted from a phyloseq object with the create_phyloseq() function.</td>
<td align="center">otu_table tax_data sample_data phy_tree</td>
<td align="center">The observation tables represent the phyloseq::otu_table(), phyloseq::tax_table(), phyloseq::sample_data(), and phyloseq::phy_tree(), data in the original phyloseq object.</td>
</tr>
<tr class="even">
<td align="center">1</td>
<td align="center">Raw Format</td>
<td align="center">A taxmap object that has been processed with the as_raw_format()function.</td>
<td align="center">phyloseq tables otu_abundance otu_annotations</td>
<td align="center">The new observation tables are just name conversions of the otu_table and tax_data table from the “phyloseq_format”.</td>
</tr>
<tr class="odd">
<td align="center">2</td>
<td align="center">BasicFormat</td>
<td align="center">A taxmap object that has been processed with the as_basic_format() function.</td>
<td align="center">raw tables taxa_abundance otu_proportions taxa_proportions</td>
<td align="center">This format is defined by observation data that has been processed with the metacoder::calc_*_() functions.</td>
</tr>
<tr class="even">
<td align="center">3</td>
<td align="center">AnalyzedFormat</td>
<td align="center">A taxmap object that has been processed with the as_analyzed_format() function..</td>
<td align="center">basic tables statistical_data stats_tax_data</td>
<td align="center">This format is defined by observation data that has been processed with the metacoder::compare_groups() function.</td>
</tr>
</tbody>
</table>
<p>Validation is performed internally by most of the preprocessing, analysis, and formatting functions, but it can also be done directly by using the *is_*_format()* functions. Furthermore, it is encouraged to explore the taxmap objects in order to view observation and taxonomy data. Other MicrobiomeR utilities include project management tools for creating and organizing output directories for plots, a set of color palette functions based on <em>grDevices::colorRampPalette()</em>, and a taxonomic data parsing function called <em>parse_taxonomy_silva_128()</em> used to import data annotated by the SILVA <span class="citation">(C et al. 2013)</span> database as a phyloseq object.</p>
</div>
</div>
<div id="acknowledgments" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgments" class="anchor"></a>Acknowledgments</h1>
<p>The data in this R package was generated by the University of Mississippi Medical Center’s Genomics Core and further analyzed using the Nephele platform from the National Institute of Allergy and Infectious Diseases (NIAID) Office of Cyber Infrastructure and Computational Biology (OCICB) in Bethesda, MD.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-silva_1">
<p>C, Quast, Pruesse E, Yilmaz P, Gerken J, Schweer T, Yarza P, Peplies J, and Glöckner FO. 2013. “The Silva Ribosomal Rna Gene Database Project: Improved Data Processing and Web-Based Tools.” <em>Nucleic Acids Research</em>, no. 41: D590–D596.</p>
</div>
<div id="ref-metacoder">
<p>Foster, Zachary, Thomas Sharpton, and Niklaus Grünwald. 2017. “Metacoder: An R Package for Visualization and Manipulation of Community Taxonomic Diversity Data.” <em>PLOS Computational Biology</em> 13 (2). Public Library of Science: 1–15. doi:<a href="https://doi.org/10.1371/journal.pcbi.1005404">10.1371/journal.pcbi.1005404</a>.</p>
</div>
<div id="ref-qiime">
<p>JG, Caporaso, Kuczynski J, Stombaugh J, Bittinger K, Bushman FD, Costello EK, Fierer N, et al. 2010. “QIIME Allows Analysis of High-Throughput Community Sequencing Data.” <em>Nature Methods</em> 7: 335–36. <a href="qiime.org" class="uri">qiime.org</a>.</p>
</div>
<div id="ref-phyloseq">
<p>McMurdie, Paul J., and Susan Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PLoS ONE</em> 8 (4): e61217. <a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="uri">http://dx.plos.org/10.1371/journal.pone.0061217</a>.</p>
</div>
<div id="ref-r-core">
<p>R Core Team. 2018. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>.</p>
</div>
<div id="ref-taxa">
<p>Zachary, Foster, Chamberlain Scott, and Grunwald Niklaus. 2018. <em>Taxa: An R Package Implementing Data Standards and Methods for Taxonomic Data.</em> <em>F1000Research</em>. Vol. 7. doi:<a href="https://doi.org/10.12688/f1000research.14013.1">10.12688/f1000research.14013.1</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#summary">Summary</a></li>
      <li><a href="#workflow">Workflow</a></li>
      <li><a href="#utilities">Utilities</a></li>
      <li><a href="#acknowledgments">Acknowledgments</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/grabear">Robert Gilmore</a>, <a href="https://www.shauritahutchins.com">Shaurita Hutchins</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
