<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Get Phyloseq Object — create_phyloseq • MicrobiomeR</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Get Phyloseq Object — create_phyloseq" />

<meta property="og:description" content="Create a phyloseq object using a biom file, phylogenetic tree file, and a
metadata file. Alternatively, an .Rdata file can be used.  For now this funciton takes
output from Qiime.  The data provided by this package was processed by the NIH's Nephele
pipeline." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeR</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/data-wrangling.html">Data Wrangling</a>
    </li>
    <li>
      <a href="../articles/filtering.html">Data Filtering</a>
    </li>
    <li>
      <a href="../articles/analysis.html">Data Analysis</a>
    </li>
    <li>
      <a href="../articles/color-palettes.html">Color Palettes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/paper.html">Publication</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="../articles/about.html">About</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vallenderlab/MicrobiomeR">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/datasnakes">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Phyloseq Object</h1>
    <small class="dont-index">Source: <a href='https://github.com/vallenderlab/MicrobiomeR/blob/master/R/phyloseq.R'><code>R/phyloseq.R</code></a></small>
    <div class="hidden name"><code>create_phyloseq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Create a phyloseq object using a biom file, phylogenetic tree file, and a
metadata file. Alternatively, an .Rdata file can be used.  For now this funciton takes
output from Qiime.  The data provided by this package was processed by the NIH's Nephele
pipeline.</p>
    
    </div>

    <pre class="usage"><span class='fu'>create_phyloseq</span>(<span class='kw'>biom_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>tree_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>metadata_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>treatment_group</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>parse_func</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>rdata_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>save_rooted_tree</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>recursive_save</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>biom_file</th>
      <td><p>A file in the <a href='http://biom-format.org/'>BIOM format</a>.  Default: NULL</p></td>
    </tr>
    <tr>
      <th>tree_file</th>
      <td><p>A phylogenetic tree file.  This function will generate a rooted
tree file if your phylogenetic tree isn't already rooted.  Default: NULL</p></td>
    </tr>
    <tr>
      <th>metadata_file</th>
      <td><p>Sample metadata in tab delimited format.  It's very important that you make
sure that you have a <strong>sample_id</strong> and <strong>TreatmentGroup</strong> column in your metadata
file.  You might otherwise run into problems.  Default: NULL</p></td>
    </tr>
    <tr>
      <th>treatment_group</th>
      <td><p>The column number or name in the metadata file that contains the
treatement group names.  Default: NULL</p></td>
    </tr>
    <tr>
      <th>parse_func</th>
      <td><p>The parse function used to parse taxonomy strings from
Greengenes or SILVA database.  Default: NULL</p></td>
    </tr>
    <tr>
      <th>rdata_file</th>
      <td><p>A .Rdata file.  Default: NULL</p></td>
    </tr>
    <tr>
      <th>save_rooted_tree</th>
      <td><p>A logical that determines weather or not the rooted tree
is saved.  Default: FALSE</p></td>
    </tr>
    <tr>
      <th>recursive_save</th>
      <td><p>If the directory doesn't exists create the parent directories
that don't exist as well.  Default: FALSE</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A phyloseq object, and a phylogenetic tree file if one does not already exist.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function heavily relys on the phyloseq package to import data into R.
It also requires you to use an absolute or relative path to your data files or for your data files
to be in the working directory.</p>
<p>It's very important that you make sure that you have a <strong>sample_id</strong> and <strong>TreatmentGroup</strong>
column in your metadata file.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://www.rdocumentation.org/packages/phyloseq/topics/import_biom'>import_biom</a></code>, <code><a href='https://www.rdocumentation.org/packages/phyloseq/topics/sample_data-methods'>phyloseq::sample_data()</a></code>, <code><a href='https://www.rdocumentation.org/packages/phyloseq/topics/phy_tree-methods'>phyloseq::phy_tree()</a></code>, <code><a href='https://www.rdocumentation.org/packages/phyloseq/topics/import_qiime_sample_data'>import_qiime_sample_data</a></code>, <code><a href='https://www.rdocumentation.org/packages/phyloseq/topics/merge_phyloseq'>merge_phyloseq</a></code></p>
<p><code><a href='https://www.rdocumentation.org/packages/ape/topics/root'>root</a></code>, <code><a href='https://www.rdocumentation.org/packages/ape/topics/root'>is.rooted</a></code>, <code><a href='https://www.rdocumentation.org/packages/ape/topics/read.tree'>read.tree</a></code>, <code><a href='https://www.rdocumentation.org/packages/ape/topics/write.tree'>write.tree</a></code></p>
<p><code><a href='root_by_longest_edge.html'>root_by_longest_edge</a></code></p>
<p>Other Data Importers: <code><a href='create_taxmap.html'>create_taxmap</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='kw'>if</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/interactive'>interactive</a></span>()){
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>MicrobiomeR</span>)
<span class='co'># Get the datafiles from the package</span>
<span class='no'>biom_file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"silva_OTU.biom"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"MicrobiomeR"</span>)
<span class='no'>tree_file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"silva.tre"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"MicrobiomeR"</span>)
<span class='no'>md_file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"nephele_metadata2.txt"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"MicrobiomeR"</span>)
<span class='no'>parse_func</span> <span class='kw'>&lt;-</span> <span class='no'>parse_parse_taxonomy_silva_128</span>
}
<span class='co'># Create a phyloseq object from the data files.</span>
<span class='no'>phy_obj</span> <span class='kw'>&lt;-</span> <span class='fu'>create_phyloseq</span>(<span class='kw'>biom_file</span> <span class='kw'>=</span> <span class='no'>biom_file</span>, <span class='kw'>tree_file</span> <span class='kw'>=</span> <span class='no'>tree_file</span>, <span class='kw'>metadata_file</span> <span class='kw'>=</span> <span class='no'>md_file</span>, <span class='kw'>parse_func</span> <span class='kw'>=</span> <span class='no'>parse_func</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/grabear'>Robert Gilmore</a>, <a href='https://www.shauritahutchins.com'>Shaurita Hutchins</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

