<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Get Phyloseq Object — get_phyloseq_obj • MicrobiomeR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Get Phyloseq Object — get_phyloseq_obj" />

<meta property="og:description" content="Create a phyloseq object using a biom file, phylogenetic tree file, and a
metadata file. Alternatively, an .Rdata file can be used.  For now this funciton takes
output from Qiime.  The data provided by this package was processed by the NIH's Nephele
pipeline." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vallenderlab/MicrobiomeR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get Phyloseq Object</h1>
    <small class="dont-index">Source: <a href='https://github.com/vallenderlab/MicrobiomeR/blob/master/R/phyloseq.R'><code>R/phyloseq.R</code></a></small>
    <div class="hidden name"><code>get_phyloseq_obj.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Create a phyloseq object using a biom file, phylogenetic tree file, and a
metadata file. Alternatively, an .Rdata file can be used.  For now this funciton takes
output from Qiime.  The data provided by this package was processed by the NIH's Nephele
pipeline.</p>
    
    </div>

    <pre class="usage"><span class='fu'>get_phyloseq_obj</span>(<span class='kw'>biom_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>tree_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>metadata_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>parse_func</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rdata_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>biom_file</th>
      <td><p>A file in the BIOM format. Default: NULL</p></td>
    </tr>
    <tr>
      <th>tree_file</th>
      <td><p>A Phylogenetic tree file. Default: NULL</p></td>
    </tr>
    <tr>
      <th>metadata_file</th>
      <td><p>Sample metadata in tab delimited format. Default: NULL</p></td>
    </tr>
    <tr>
      <th>parse_func</th>
      <td><p>The parse function used to parse taxonomy strings from
Greengenes or SILVA database. Default: NULL</p></td>
    </tr>
    <tr>
      <th>rdata_file</th>
      <td><p>A .Rdata file.  Default: NULL</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>A path for the output tree file.  Default: NULL</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A phyloseq object, and a phylogenetic tree file if one does not already exist.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function heavily relys on the phyloseq package to import data into R.
It also requires you to use an absolute path to your data files or for your data files
to be in the working directory.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/import_biom'>import_biom</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/phy_tree-methods'>phy_tree</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/import_qiime_sample_data'>import_qiime_sample_data</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/merge_phyloseq'>merge_phyloseq</a></code>
<code><a href='http://www.rdocumentation.org/packages/ape/topics/root'>root</a></code>
<code><a href='root_by_longest_edge.html'>root_by_longest_edge</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"># NOT RUN {
> biom_file <- "input_data/silva_OTU.biom"
> md_file <- "input_data/nephele_metadata.txt"
> phy_obj <- get_phyloseq_obj(biom_file=biom_file, metadata_file=md_file)
# }
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Robert Gilmore, Shaurita Hutchins.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

