<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Preprocessing for Phyloseq Objects — preprocess_phyloseq • MicrobiomeR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Preprocessing for Phyloseq Objects — preprocess_phyloseq" />

<meta property="og:description" content="This function uses techniques from the Bioconductor Workflow in order to
preprocess phyloseq data for downstream analysis." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/vallenderlab/MicrobiomeR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Preprocessing for Phyloseq Objects</h1>
    <small class="dont-index">Source: <a href='https://github.com/vallenderlab/MicrobiomeR/blob/master/R/phyloseq.R'><code>R/phyloseq.R</code></a></small>
    <div class="hidden name"><code>preprocess_phyloseq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function uses techniques from the Bioconductor Workflow in order to
preprocess phyloseq data for downstream analysis.</p>
    
    </div>

    <pre class="usage"><span class='fu'>preprocess_phyloseq</span>(<span class='no'>phyloseq_object</span>, <span class='kw'>process_list</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>phyloseq_object</th>
      <td><p>A phyloseq object</p></td>
    </tr>
    <tr>
      <th>process_list</th>
      <td><p>This parameter is used to control the way the phyloseq object is
processed.  It can be one of three values:</p><dl class='dl-horizontal'>
<dt>NULL</dt><dd><p>Will utilize the default processing strategy or the dot parameters (...).</p></dd>
<dt>file</dt><dd><p>The absolute path to a YAML config file.</p></dd>
<dt>list</dt><dd><p>Equivalent to putting dot parameters in a list (list(...)).</p></dd>
</dl></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>The dot parameters can be any combination of the default processing keywords.
See the "Keywords for Processing" section below for more details</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a phyloseq object that has undergone the specified processing strategies.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>DETAILS</p>
    
    <h2 class="hasAnchor" id="keywords-for-pre-processing"><a class="anchor" href="#keywords-for-pre-processing"></a>Keywords for Pre-processing</h2>

    
    <dl class='dl-horizontal'>
<dt>master_thresh</dt><dd><p>DEFAULT: 1e-5.  Filters any taxa that do not meat the mean threshold.</p></dd>
<dt>taxon_filter</dt><dd><p>DEFAULT: list("Phylum"= list("min_a"=5, "r_s_p"=0.5),
"Class"=list("min_a"=3, "r_s_p"=0.3)).  Filters OTUs that do not appear more than a certian
amount of times in a certain percentage of samples at the specified agglomerated rank.</p></dd>
<dt>prevelance_filter</dt><dd><p>DEFAULT: list("min_a"=5, "r_s_p"=0.5).  Filters OTUs that do not
appear more than a certian amount of times in a certain percentage of samples.</p></dd>
<dt>glom_rank</dt><dd><p>DEFAULT: NULL.  Agglomerates the data at the specified rank.</p></dd>
<dt>ambiguous</dt><dd><p>DEFAULT: list(amb_ranks = c("Phylum", "Class", "Order", "Family", "Genus"),
amb_items = c(NA, "", "uncharacterized", "uncultured", "Unassigned", "Ambiguous", "Ambiguous_taxa")).
Removes OTUs that are labeled with the specified ambiguous items.  This is done for each specified rank.</p></dd>
<dt>coeff_of_variation</dt><dd><p>DEFAULT: 0.55.  Standardizes abundances to the median sequencing depth</p></dd>
<dt>trans_function</dt><dd><p>DEFAULT:  function(x)x / sum(x).  Transforms the abundance values to relative
abundance values.</p></dd>
<dt>merge_samples</dt><dd><p>DEFAULT:  NULL.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/prune_taxa-methods'>prune_taxa</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/taxa_sums'>taxa_sums</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/filter_taxa'>filter_taxa</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/tax_glom'>tax_glom</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/nsamples-methods'>nsamples</a></code>,
<code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/filterfun_sample'>filterfun_sample</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/genefilter_sample-methods'>genefilter_sample</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/get_taxa_unique'>get_taxa_unique</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/transformcounts'>transform_sample_counts</a></code>,
<code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/merge_samples-methods'>merge_samples</a></code>, <code><a href='http://www.rdocumentation.org/packages/phyloseq/topics/subset_taxa-methods'>subset_taxa</a></code></p>
<p><code><a href='http://www.rdocumentation.org/packages/yaml/topics/yaml.load'>yaml.load</a></code><br /><br />
Bioconductor Workflow - <a href='https://f1000research.com/articles/5-1492/v2'>https://f1000research.com/articles/5-1492/v2</a><br />
Phyloseq Website - <a href='https://joey711.github.io/phyloseq/index.html'>https://joey711.github.io/phyloseq/index.html</a></p>
<p>Other Filters: <code><a href='remove_ambiguous_taxa.html'>remove_ambiguous_taxa</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"># NOT RUN {
> phy_obj <- get_phyloseq_object(...)
> phy_obj
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 1955 taxa and 48 samples ]
sample_data() Sample Data:       [ 48 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 1955 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 1955 tips and 1954 internal nodes ]
> pp_phy_obj <- preprocess_phyloseq(phy_obj, master_thresh = 1e-5,
                                    taxon_filter = list("Phylum"= list("min_a"=5, "r_s_p"=0.5),
                                                        "Class"=list("min_a"=3, "r_s_p"=0.3)),
                                    prevalence_filter = list("min_a"=5, "r_s_p"=0.5), glom_rank = NULL,
                                    ambiguous=list(amb_ranks = c("Phylum", "Class", "Order", "Family", "Genus"),
                                    amb_items = c(NA, "", "uncharacterized", "uncultured", "Unassigned", "Ambiguous", "Ambiguous_taxa")),
                                    coeff_of_variation = 0.55, trans_function = function(x){x / sum(x)}, merge_samp = NULL)
> pp_phy_obj

# }
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#keywords-for-pre-processing">Keywords for Pre-processing</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Robert Gilmore, Shaurita Hutchins.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

